README.rst
The Whatever Object
===================

An easy way to make lambdas by partial application of python operators.

Inspired by Perl 6 one, see http://perlcabal.org/syn/S02.html#The_Whatever_Object


Usage
-----

.. code:: python

    from whatever import _, that

    # get a list of guys names
    names = map(_.name, guys)
    names = map(that.name, guys)

    odd = map(_ * 2 + 1, range(10))

    squares = map(_ ** 2, range(100))
    small_squares = filter(_ < 100, squares)

    best = max(tries, key=_.score)
    sort(guys, key=-that.height)

    factorial = lambda n: reduce(_ * _, range(2, n+1))

NOTE: chained comparisons cannot be implemented since there is no boolean overloading in python.


CAVEATS
-------

In some special cases whatever can cause confusion:

.. code:: python

    _.attr # this makes callable
    obj._  # this fetches '_' attribute of obj

    _[key] # this works too
    d[_]   # KeyError, most probably

    _._    # short for attrgetter('_')
    _[_]   # short for lambda d, k: d[k]

    if _ == 'Any value':
        # You will get here, definitely
        # `_ == something` produces callable, which is true

    [1, 2, _ * 2, None].index('hi') # => 2, since bool(_ * 2 == 'hi') is True


Also, whatever sometimes fails on late binding:

.. code:: python

    (_ * 2)('2') # -> NotImplemented
    
Additionally ``whatever._`` is incompatible with IPython's output caching
(enabled by default) which uses ``_`` to store the previous result, see https://ipython.org/ipython-doc/dev/interactive/reference.html#output-caching-system.


CHANGELOG
0.7
- support Python 3.11
- include CHANGELOG into a release (Igor Gnatenko)

0.6
- support Python 3.8 (Felix Yan)
- explicit support for Python 3.7
Backward incompatible changes:
- dropped Python 2.6 and 3.3 support

0.5
- added caller functionality to whatever
- improved __code__ introspection
- optimized things a bit

0.4.3
- explicit support for Python 3.6
- minor optimizations

0.4.2
- explicit support for Python 3.5 (Felix Yan)

0.4.1
- support pypy3 and python 3.4
- document late binding not supported

0.4
- significant optimizations
- support arity introspection for unary ops
Backward incompatible changes:
- previous clumsy support for overloadded ops dropped
- WhateverCode is now true in boolean context

0.3
- added Python 3 support (ashleyh)

0.2.2
...


test_whatever.py
import pytest
from whatever import _, that


def test_basic():
    assert _ is that


def test_caller():
    assert _(11)(lambda x: x ** 2) == 121
    assert _('100', base=11)(int) == 121


def test_attrs():
    assert callable(_.attr)

    class A(object):
        def __init__(self, name):
            self.name = name

    assert _.name(A('zebra')) == 'zebra'

    class B(object):
        def __getattr__(self, name):
            return name * 2

    b = B()
    assert _.foo(b) == 'foofoo'


def test_getitem():
    assert callable(_[0])
    assert _[0]([1, 2, 3]) == 1
    assert _[1:]([1, 2, 3]) == [2, 3]

    assert callable(_['item'])
    assert _['name']({'name': 'zebra'}) == 'zebra'


def test_add():
    # last and right
    assert (_ + 1)(10) == 11
    assert (1 + _)(10) == 11

    # non commutative
    assert (_ + 'y')('x') == 'xy'
    assert ('y' + _)('x') == 'yx'


def test_late_binding():
    assert (_ * 2)(2) == 4
    assert (_ * 2)('2') is NotImplemented


@pytest.mark.xfail
def test_overloaded():
    class StickyString(str):
        def __add__(self, other):
            if other is _:
                return NotImplemented
            return str(self) + str(other)

        def __radd__(self, other):
            if other is _:
                return NotImplemented
            return str(other) + str(self)

    foo = StickyString('foo')
    assert foo + 2 == 'foo2'
    assert (_ + 2)(foo) == 'foo2'
    assert (foo + _)(2) == 'foo2'
    assert (_ + foo)(2) == '2foo'


def test_comparison():
    assert callable(_ < 10)
    assert (_ < 10)(5) is (5 < 10)
    assert (_ < 10)(15) is (15 < 10)

    assert (_ == None)(None) is (None == None)
    assert (_ != None)(None) is (None != None)

    assert list(filter(5 < _, range(10))) == [6, 7, 8, 9]


def test_unary():
    assert callable(-_)
    assert callable(abs(_))

    assert (-_)(5) == -5
    assert abs(_)(-5) == 5
    assert min([2,3,5,6], key=-_) == 6


def test_chained_ops():
    assert callable(_ + 1 + 2)
    assert (_ + 1 + 2)(10) == 13

    assert (_ * 2 + 3)(1) == 5
    assert (_ + 2 * 3)(1) == 7

    assert (_ + 1 + 1 + 1 + 1 + 1)(0) == 5
    assert (_ + 1 + 1)(1) == 3
    assert (1 + _ + 1)(1) == 3
    assert (1 + (1 + _))(1) == 3

    assert ( _  + 'y' + 'z')('x') == 'xyz'
    assert ('x' +  _  + 'z')('y') == 'xyz'
    assert ('x' + 'y' +  _ )('z') == 'xyz'

    assert -abs(_)(-5) == -5
    assert abs(3 - _)(10) == 7


# NOTE: this is impossible since there is no boolean overloading in python
# def test_chained_comparison():
#     assert (1 < _ < 10)(7)
#     assert not (1 < _ < 10)(10)
#     assert not (1 < _ < 10)(1)


def test_chained_attrs():
    class A(object):
        def __init__(self, val):
            self.val = val

    assert callable(-that.val)
    assert (-that.val)(A(10)) == -10
    assert (that.val + 100)(A(10)) == 110

    assert ((_ + ', ' + 'Guys!').lower)('Hi')() == 'hi, guys!'
    assert that.val.val(A(A('a value'))) == 'a value'


def test_chained_getitem():
    assert callable(-that[0])
    assert callable(-that[2:])
    assert callable(that['key'] + 1)
    assert callable((_ + 1)['key'])

    assert (-_[0])([1,2,3]) == -1
    assert (_[1:] + [4])([1,2,3]) == [2,3,4]
    assert (_ + [4])[2:]([1,2,3]) == [3,4]

    assert (_['val'] * 5)({'val': 2}) == 10
    assert _['i']['j']({'i': {'j': 7}}) == 7


def test_higher_cardinality():
    assert (_ + _)(1, 2) == 3
    assert (_ + _ + 1)(1, 2) == 4
    assert (1 + (_ + _))(1, 2) == 4
    assert (_ ** _ ** _)(2, 3, 4) == 2 ** 3 ** 4
    assert ((_ ** _) ** _)(2, 3, 4) == (2 ** 3) ** 4
    assert (_ ** (_ ** _))(2, 3, 4) == 2 ** (3 ** 4)


def test_introspect_whatever():
    code = _.__code__
    assert code.co_argcount == 1
    assert len(code.co_varnames) == 1
    assert code.co_nlocals == 1


def test_introspection():
    assert (_ + 1).__code__.co_argcount == 1
    assert (_ + _).__code__.co_argcount == 2
    assert (_ + 1 + _).__code__.co_argcount == 2

    code = (_ + _).__code__
    assert 'add' in code.co_name
    assert len(code.co_varnames) == 2
    assert code.co_nlocals == 2


def test_code_to_code():
    assert (_ ** 2 + _ ** 2)(3, 4) == 5 ** 2


def test_contains():
    with pytest.raises(NotImplementedError): 1 in _


whatever.py
# -*- coding: utf-8 -*-
"""
The Whatever object inspired by Perl 6 one.

See http://perlcabal.org/syn/S02.html#The_Whatever_Object
"""
import operator, sys
from types import CodeType

__all__ = ['_', 'that']
PY2 = sys.version_info[0] == 2
PY38 = sys.version_info[:2] >= (3, 8)
PY311 = sys.version_info[:2] >= (3, 11)


# TODO: or not to do
# object.__call__(self[, args...])

# TODO: pow with module arg:
# object.__pow__(self, other[, modulo])


def _make_code(arity, name, varnames):
    # Python 3 got kwonlyargcount, Python 3.8 got posonlyargcount
    args = [arity, arity] if PY2 else [arity, 0, 0, arity] if PY38 else [arity, 0, arity]
    args.extend([1, 67, b'', (), (), varnames, __name__, name])
    # Python 3.11 got qualname
    if PY311:
        args.append('%s.%s' % (__name__, name))
    args.extend([1, b''])
    # Python 3.11 got exceptiontable
    if PY311:
        args.append(b'')
    return CodeType(*args)


class Whatever(object):
    def __contains__(self, other):
        raise NotImplementedError('Sorry, can\'t to hook "in" operator in this way')

    @staticmethod
    def __call__(*args, **kwargs):
        return WhateverCode.make_call(lambda f: f(*args, **kwargs), 1)

    __code__ = _make_code(1, 'Whatever', ('f',))


class WhateverCode(object):
    def __init__(self, arity):
        self._arity = arity

    @classmethod
    def make_call(cls, func, arity):
        sub_cls = type('WhateverCodeCall', (WhateverCode,), {'__call__': staticmethod(func)})
        return sub_cls(arity)

    def __contains__(self, other):
        raise NotImplementedError('Sorry, can\'t hook "in" operator in this way')

    # Simulate normal callable
    @property
    def __code__(self):
        fname = self.__call__.__name__ or 'operator'
        varnames = tuple('abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ'[:self._arity])
        return _make_code(self._arity, fname, varnames)


### Unary ops

def unary(op):
    return lambda self: WhateverCode.make_call(op, 1)

def code_unary(op):
    return lambda self: WhateverCode.make_call(lambda that: op(self(that)), self._arity)


### Binary ops

def operand_type(value):
    return Whatever     if isinstance(value, Whatever) else     \
           WhateverCode if isinstance(value, WhateverCode) else \
           None

def argcount(operand):
    op_type = operand_type(operand)
    if op_type is None:
        return 0
    elif op_type is Whatever:
        return 1
    else:
        return operand._arity

def gen_binary(op, left, right):
    W, C, D = Whatever, WhateverCode, None
    name, rname, func, args = op
    ltype, rtype = types = operand_type(left), operand_type(right)

    # Constant incorporating optimizations
    lfunc = rfunc = None
    if ltype is D:
        _lfunc = lambda x: func(left, x)
        lfunc = getattr(left, name, _lfunc) if name else _lfunc
    if rtype is D:
        if name == '__getattr__':
            assert isinstance(right, str)
            # NOTE: eval('lambda x: x.%s' % right) is even faster, but to slow to construct
            rfunc = operator.attrgetter(right)
        elif name == '__getitem__':
            rfunc = operator.itemgetter(right)
        else:
            _rfunc = lambda x: func(x, right)
            rfunc = getattr(right, rname, _rfunc) if rname else _rfunc

    # Resolve embedded __call__ in advance
    if ltype is C:
        lcall = left.__call__
    if rtype is C:
        rcall = right.__call__
    largs, rargs = argcount(left), argcount(right)

    ops = {
        (W, D): rfunc,
        (D, W): lfunc,
        # (C, D) are optimized for one argument variant
        (C, D): (lambda x: rfunc(lcall(x))) if largs == 1 else
                        (lambda *xs: rfunc(lcall(*xs))),
        (D, C): (lambda x: lfunc(rcall(x))) if rargs == 1 else
                        (lambda *xs: lfunc(rcall(*xs))),
        (W, W): func,
        (W, C): lambda x, *ys: func(x, rcall(*ys)),
        (C, W): lambda *xs: func(lcall(*xs[:-1]), xs[-1]),
        (C, C): lambda *xs: func(lcall(*xs[:largs]), rcall(*xs[largs:])),
    }
    return WhateverCode.make_call(ops[types], arity=largs + rargs)

def binary(op):
    return lambda left, right: gen_binary(op, left, right)

def rbinary(op):
    return lambda left, right: gen_binary(op, right, left)


### Define ops

def op(name, rname=None, func=None, args=2):
    name = '__%s__' % name
    if rname:
        rname = '__%s__' % rname
    return (name, rname, func or getattr(operator, name), args)

def rop(name, func=None):
    return op(name, 'r' + name, func=func, args=2)

def ops(names, args=2):
    return [op(name, args=args) for name in names]

def rops(names):
    return [rop(name) for name in names]


OPS = rops(['add', 'sub', 'mul', 'floordiv', 'truediv', 'mod', 'pow',
            'lshift', 'rshift', 'and', 'xor', 'or'])                       \
    + [rop('divmod', func=divmod)]                                         \
    + ops(['eq', 'ne', 'lt', 'le', 'gt', 'ge'])                            \
    + ops(['neg', 'pos', 'abs', 'invert'], args=1)                         \
    + [op('getattr', func=getattr), op('getitem')]

# This things were dropped in python 3
if PY2:
    OPS += [rop('div'), op('cmp', func=cmp)]  # noqa

for op in OPS:
    name, rname, func, args = op
    if args == 1:
        setattr(Whatever, name, unary(func))
        setattr(WhateverCode, name, code_unary(func))
    elif args == 2:
        setattr(Whatever, name, binary(op))
        setattr(WhateverCode, name, binary(op))
        if rname:
            setattr(Whatever, rname, rbinary(op))
            setattr(WhateverCode, rname, rbinary(op))

_ = that = Whatever()
